Add setting for handling wildcard mentions in large streams.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
We add a setting for handling wildcard mentions in large streams (i.e. streams with more than 15 subscribers).
The setting has following options - 
1. allow everyone to use wildcard mention.
2. Show warning when using wildcard mention.
3. Allow only admins to use wildcard mention.

Fixes #16211.

**Testing Plan:** <!-- How have you tested? --> Added automated tests and manually tested on dev server.


 <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on zerver/lib/actions.py:
```diff
@@ -2299,6 +2295,11 @@ def check_message(sender: UserProfile, client: Client, addressee: Addressee,
+        content=message_content
+    )
+
     if addressee.is_stream():
```
I have made this change to extract `MentionData` in `check_message` instead of `do_send_messages` because I think that it would be better to check for the message here itself instead of doing it in `do_send_messages`, though this leads to some more changes like in management commands where we send messages directly calling `do_send_messages` and not call `check_message`, (Just mentioning it as I am not fully sure that whether this is the best approach or not.)

showell commented on zerver/tests/test_message_send.py:
```diff
@@ -1321,6 +1321,68 @@ def test_message_bot_mentions(self) -> None:
+        self.assertEqual(last_message.content, content)
+
+        do_set_realm_property(cordelia.realm, 'wildcard_mention_policy', Realm.WILDCARD_MENTION_ALLOW_EVERYONE)
+        result = self.client_post("/json/messages", {"type": "stream",
```
It might be worth running this new chunk of code through Black, as we're probably eventually gonna go away from the hanging-indent style.

sahil839 replied:
But I am not sure how to run this code through Black. Should it be done from commits from #15662?

showell replied:
Install black (google “Python black code formatter” or similar).  Copy the big chunk of code here to temp.py and run black against it.  Then paste back into the existing file.  Make sense?

sahil839 replied:
Yeah got it.

sahil839 replied:
Updated the code with Black formatting.

showell commented on static/js/server_events_dispatch.js:
```diff
@@ -143,6 +143,7 @@ exports.dispatch_normal_event = function dispatch_normal_event(event) {
                 emails_restricted_to_domains: noop,
                 video_chat_provider: compose.update_video_chat_button_display,
                 waiting_period_threshold: noop,
+                wildcard_mention_policy: noop,
```
This is unrelated to your change, but I wonder why we need all these `noop` lines.  I guess explicit is better, but if the most common case is to to do no extra processing, I wonder why we don't just change the code to skip the post-processing step if the field's not present in this dict.

sahil839 commented:
@timabbott This PR is ready for review.

sahil839 commented:
Just updated the commit message in last push.

timabbott commented:
@sahil839 thanks for working on this!  I think this logic as written will end up firing on code blocks containing wildcard mention syntax; I expect we'll need a more complex refactoring to make this work as intended.  I'll post some comments; it's possible you should do that preparatory refactor in a separate PR for convenience of review/discussion.

timabbott commented on zerver/management/commands/deliver_scheduled_messages.py:
```diff
@@ -49,6 +50,8 @@ def construct_message(self, scheduled_message: ScheduledMessage) -> Dict[str, An
             content=scheduled_message.content,
         )
+        if scheduled_message.stream is not None and mention_data.message_has_wildcards():
+            check_wildcard_mention_restrictions(original_sender, scheduled_message.stream)
```
If you look at `zerver/lib/mention.py`, you'll see that `possible_mentions` (which populates MentionData) just looks for mention-like syntax; it doesn't check for actual mentions.  That latter bit can only be done with actual markdown rendering.

I think the right fix is that we should move the `render_incoming_message` call from `do_send_messages` to `check_message`, so that we can access the actual mention details inside `check_message`.

@showell FYI in case you want to pick up this refactor and/or collaborate on it.

sahil839 commented:
Opened #16422 for refactoring.

sahil839 commented:
@timabbott I have updated the PR to not allow sending messages when wildcard mention syntax is present in code block using the API, but for frontend part I am not able to figure out how to do this and it still shows warning/error.
I have checked the current behaviour on master where we only show warnings for wildcard mentions, the warning is also shown for mention syntax present in code blocks, as we only check it using `message_content.match(/(^|\s)(@\*{2}(all|everyone|stream)\*{2})($|\s)/);` and this line does not check whether it is in code block or not. I am not able to figure out how to fix this.

timabbott commented on frontend_tests/puppeteer_tests/09-mention.js:
```diff
@@ -19,8 +19,8 @@ async function test_mention(page) {
-        compose.all_everyone_warn_threshold = 5;
-        return compose.all_everyone_warn_threshold;
+        compose.all_everyone_mention_threshold = 5;
+        return compose.all_everyone_mention_threshold;
```
I renamed this to `wildcard_mention_large_stream_threshold` for better readability.

timabbott commented on zerver/lib/message.py:
```diff
@@ -1212,3 +1215,35 @@ def get_recent_private_conversations(user_profile: UserProfile) -> Dict[int, Dic
+    if realm.wildcard_mention_policy == Realm.WILDCARD_MENTION_POLICY_MEMBERS:
+        return not sender.is_guest
+
+    raise AssertionError("Invalid wildcard mention policy")
```
I rewrote this function to have a cleaner style and support the full set of policy options we'd like to have.

timabbott commented on zerver/tests/test_message_send.py:
```diff
@@ -1317,6 +1317,108 @@ def test_message_bot_mentions(self) -> None:
+        )
+        self.assert_json_success(result)
+        last_message = self.get_last_message()
+        self.assertEqual(last_message.content, content)
```
We should extract this block as a helper function; probably `self.send_stream_message()` to save some duplicated code, and maybe a small wrapper for this specific test.  I'm not sure we need to do the `last_message` / `content` verification.

timabbott replied:
(And we should test all 6 settings values).

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2401,7 +2402,12 @@ def check_message(sender: UserProfile, client: Client, addressee: Addressee,
+
+    if stream is not None and message_send_dict['message'].mentions_wildcard:
+        if not wildcard_mention_allowed(sender, stream):
+            raise JsonableError_("You do not have permission to use wildcard mentions here.")
```
Maybe should be "mentions in this stream" rather than "... here"?

timabbott commented:
I merged the first few commits as the series ending with 2e58e27926a82629ee6ef505becec9f8c9658004, after reworking the set of policy options to allow, and pushed back a new version of the enforcement logic appropriate for the new set of options.  Thanks for all your work on this so far @sahil839!

I think we just need you to redo the tests and other details that my changes broke, and we should be ready to integrate this! :)

timabbott commented:
On the question of the warning logic, I think we want to show the warning unconditionally to anyone who is in fact allowed to send, since it's basically a reminder that the operation is disruptive (just like deleting messages, etc.).  So we don't need that to be considered in the setting (which really simplifies the problem statement here and is what made providing a wide range of options for which users can do things make sense).

sahil839 commented:
I will make the suggested changes by tomorrow . 
Just wanted to clear my above question- In the current implementation of master in front-end, we also show warning if wildcard mention syntax is inside the codeblocks, which I think we do not want.
The code which Checks whether there is wildcard mention or not is 
`message_content.match(/(^|\s)(@\*{2}(all|everyone|stream)\*{2})($|\s)/)`. I am not able to figure out how to fix this such that we do not consider them as wildcard mention if the syntax for wildcard mention is present in code blocks.

timabbott commented:
I don't think we can readily fix the frontend notice being potentially in error without doing markdown rendering to confirm.  Let's leave that for a separate issue -- it's basically unrelated and I don't think it's likely to be an issue in practice.

timabbott commented:
@sahil839 do you have a chance to update this?  I think it's really close and it'd be nice to clear out the failure on the chat.zulip.org branch.

sahil839 commented:
@timabbott This PR is ready for review, have done the following changes -
- Added commit for modifying realm tests for considering all values of wildcard mention policy setting.
- Added commit for updating the default value from stream admins to realm admins as we have not added stream admins options in frontend yet.
- Modified the wildcard_mention_allowed code added by you to allow admins to use wildcard mentions even when they are new and wildcard mention policy is set to FULL_MEMBERS_ONLY as we do for other policy with similar options.
- Modified frontend commit for setting to consider all values of wildcard mention policy and have commented out stream admins option for now.
- Added a function `wildcard_mention_allowed` in frontend to accordingly show error message in frontend. For now we are showing the same error message for all cases as we do in the backend.

timabbott commented on zerver/models.py:
```diff
@@ -268,7 +268,7 @@ class Realm(models.Model):
     WILDCARD_MENTION_POLICY_NOBODY = 6
     wildcard_mention_policy: int = models.PositiveSmallIntegerField(
-        default=WILDCARD_MENTION_POLICY_STREAM_ADMINS,
+        default=WILDCARD_MENTION_POLICY_ADMINS,
```
I'm not sure we need to do this -- it might be cleaner to do a hack in how the feature is described to users for the intermediate period.

sahil839 replied:
Also, I was thinking whether it will be better to keep ADMINS only as default instead of STREAM_ADMINS, as per our current approach, we would make the user creating the stream a stream admin, so such users would also be allowed to use wildcard mentions in those streams.

timabbott replied:
Yeah, I was actually just thinking about that question.  I think that the right way to stop that sort of abuse is to restrict who can create streams (which we recommend in our "moderating open organizations" document) for that sort of public community.

(Creating a stream and then subscribing other users to it is disruptive in the same way that a wildcard mention is).

timabbott commented on static/js/compose.js:
```diff
@@ -500,6 +538,12 @@ function validate_stream_message_mentions(stream_id) {
+                i18n.t("You do not have permission to use wildcard mentions in this stream."),
+            );
+            return false;
+        }
```
I think this check would be better as the first one in this block -- it lets us avoid calling `wildcard_mention_allowed` twice.  I took care of moving this.

timabbott commented on static/js/settings_config.js:
```diff
@@ -137,6 +137,40 @@ exports.private_message_policy_values = {
+    //     order: 4,
+    //     code: 4,
+    //     description: i18n.t("Organization and stream admins"),
+    // },
```
I replaced this with a different hack that allows us to keep the default value as 4, which will eventually mean "Organization and stream admins", since that's cleaner than doing database migrations back and forth.

Hopefully we can merge the pieces of "stream admins" needed to fully implement this feature soon.

timabbott commented:
Merged as the series ending with 465122b0dffddddb325035ea294df6d580947f48, after doing the above and fixing a testing bug I'll post about shortly.

timabbott commented on zerver/tests/test_message_send.py:
```diff
@@ -1325,6 +1325,104 @@ def test_message_bot_mentions(self) -> None:
+        )
+        do_set_realm_property(realm, "waiting_period_threshold", 10)
+        iago.date_joined = timezone_now()
+        cordelia.date_joined = timezone_now()
```
We need to save iago/cordelia here; otherwise, this test will start failing on any local database where the users are more than 10 days old :)

timabbott commented:
Thanks for doing this @sahil839!  I know this PR ended up being a more complicated chain of work than we'd expected (especially with the need to refactor `do_send_messages` to be able to write the required check).

I think once we have integrated enough of the "stream admins" PR to have the `role` field available to the frontend, we should do a separate PR to clear out the TODOs created by this PR -- even if there's no UI for stream admins, we can write and test the logic for it once we have a place to put the values.

Can you open a `zulip-mobile` issue to make sure they're aware of this new feature?

sahil839 commented:
Opened [#4286](https://github.com/zulip/zulip-mobile/issues/4286). 

One more thing, we should probably also add a separate section for this setting in /help page.
